python-ffi/Data/Array.py

```diff
@@ -137,24 +142,24 @@ def _updateAtImpl(just, nothing, i, a, l):
         return nothing
     ll = list(l)
     ll[i] = a
-    return just(ll)
+    return just(tuple(ll))
 
 
 _updateAt = lambda just: lambda nothing: lambda i: lambda a: lambda l: _updateAtImpl(
     just, nothing, i, a, l
 )
 
-reverse = lambda xs: list(reversed(xs))
+reverse = lambda xs: tuple(reversed(xs))
 
 
 def concat(xss):
     result = []
     for x in xss:
         result.extend(x)
```

Comment:
bind `extend = result.extend` outside the loop.
There're a lot of same places you'd follow my above review to change